code/contracts/ONEWallet.sol

```diff
@@ -64,30 +65,33 @@ contract ONEWallet {
         (uint32 ct, bool completed) = _findCommit(hash);
         require(ct == 0 && !completed, "Commit already exists");
         Commit memory nc = Commit(hash, uint32(block.timestamp), false);
-        commits.push(nc);
+        commits.push(nc); // IH: would not it be better to have mapping of commits?
     }
 
     // TODO (@polymorpher): during reveal, reject if index corresponds to a timestamp that is same as current block.timestamp, so we don't let the client accidentally leak EOTP that is still valid at the current time, which an attacker could use to commit and reveal a new transaction. This introduces a limitation that there would be a OTP_INTERVAL (=30 seconds by default) delay between commit and reveal, which translates to an unpleasant user experience. To fix this, we may slice OTP_INTERVAL further and use multiple EOTP per OTP_INTERVAL, and index the operations within an interval by an operation id. For a slice of size 16, we will need to generate a Merkle Tree 16 times as large. In this case, we would reject transactions only if both (1) the index corresponds to a timestamp that is same as current block.timestamp, and, (2) the operation id is not less than the id corresponding to the current timestamp's slice
+    // IH: what do you state above just supports HOTPs instead of TOTPs, besides other advantages.
```

Comment:
We also discussed this in the chat. As I mentioned

> By the way this TODO was already done. Forgot to remove the comment.

Nonetheless, I agree switching to HOTP would eliminate this problem, since we would allow each OTP (and associated proof) to be used exactly once. But the nonce-mechanism implemented here also achieves the same result. My main concern for using a HOTP-based mechanism in the single-factor setting is the following: 

(1) since we have to allow synchronization of counters between smart contract, authenticator, and client, we would also have to always accept HOTPs (and their proofs) and later indices. This means the permission to operate the smart contract wallet is not evenly divided into time-slices, as we would have nicely in the current TOTP-based mechanism. Instead, each index would essentially give a permission to operate the wallet once, but once a later index is used, it would invalidate an earlier index, rendering the earlier index useless.

(2) The client needs to load all hashed, salted HOTPs in proximity to the expected counter, to effectively synchronize the counter with the authenticator. In TOTP based mechanism, the client only needs to load one or two by pinpointing the current time. The difference is not significant at present, but consider in the future we may store the hashed-salted OTPs in a more secure storage, the difference would mean that the former approach (HOTP) would be more vulnerable at the client.

(3) HOTP is much less commonly used. People are much less used to operate on HOTP and the chance of human error is not low. On Google Authenticator, once the current HOTP is displayed, it stays displayed for a long time unless the user switches out from the app, or manually click the refresh button. This would be problematic if the user is using the wallet on the computer, and using the authenticator on the phone. It is likely that the user cannot remember whether the current displayed code is used or not (keep in mind the user would have many other OTPs displayed on the same UI). The user would type in a code that is already used, only to find out it is not accepted, then the user would hit the refresh button and try again. This introduces non-trivial friction and annoyance which should be avoided.

Reply:
On (1), keep in mind we are still limited by the (Google) authenticator, which does not take variable input when it is generating code. The user may randomly refresh (hence increment counters) on the authenticator, hence introducing the out-of-sync problem, and force us to allow skipping to later EOTP (thus attacker only need to crack a later one, rather than being limited by a time window)

On (2) - you would be correct if we consider client as a whole - i.e. either all information at the client is exposed, or nothing is exposed. I think there should exist mechanisms which we can store majority of the information (e.g. leaves not relevant to current time) in a secure storage (e.g. a vault or enclave), and only load as minimally as possible for whatever is needed as the user executes an operation. The mechanism is still theoretical in browser, but it is quite practical if we venture into the realm of native applications (on either desktop or mobile).